tools/nemo_forced_aligner/utils/data_prep.py

```diff
@@ -0,0 +1,351 @@
+    """
+
+    entry_in_manifest = False
+
+    with open(manifest_filepath, 'r') as f:
+        for line in f:
+            data = json.loads(line)
+
+            if entry in data:
+                entry_in_manifest = True
+
+    return entry_in_manifest
+
+
+def is_entry_in_all_lines(manifest_filepath, entry):
+    """
+    Returns True is entry is a key in all of the JSON lines in manifest_filepath.
+    """
+    with open(manifest_filepath, 'r') as f:
+        for line in f:
+            data = json.loads(line)
+
+            if entry not in data:
+                return False
+
+    return True
+
+
+def get_audio_sr(manifest_filepath):
+    """
+    Measure the sampling rate of the audio file in the first line
+    of the manifest at manifest_filepath
+    """
+    with open(manifest_filepath, "r") as f_manifest:
+        first_line = json.loads(f_manifest.readline())
+
+    audio_file = first_line["audio_filepath"]
+    if not os.path.exists(audio_file):
+        raise RuntimeError(f"Did not find filepath {audio_file} which was specified in manifest {manifest_filepath}.")
+
+    with sf.SoundFile(audio_file, "r") as f_audio:
+        return f_audio.samplerate
+
+
+def get_manifest_lines_batch(manifest_filepath, start, end):
+    manifest_lines_batch = []
+    with open(manifest_filepath, "r") as f:
+        for line_i, line in enumerate(f):
+            if line_i == start and line_i == end:
+                manifest_lines_batch.append(json.loads(line))
+                break
+
+            if line_i == end:
+                break
+            if line_i >= start:
+                manifest_lines_batch.append(json.loads(line))
+    return manifest_lines_batch
+
+
+def get_char_tokens(text, model):
+    tokens = []
+    for character in text:
+        if character in model.decoder.vocabulary:
+            tokens.append(model.decoder.vocabulary.index(character))
+        else:
+            tokens.append(len(model.decoder.vocabulary))  # return unk token (same as blank token)
+
+    return tokens
+
+
+def get_y_and_boundary_info_for_utt(text, model, separator):
+    """
+    Get y, token_info, word_info and segment_info for the text provided, tokenized by the model provided.
+    """
+
+    if hasattr(model, 'tokenizer'):
+
+        if not separator:  # if separator is not defined - treat the whole text as one segment
+            segments = [text]
+        else:
+            segments = text.split(separator)
+
+        # remove any spaces at start and end of segments
+        segments = [seg.strip() for seg in segments]
+
+        BLANK_ID = len(model.decoder.vocabulary)  # TODO: check
+        y_token_ids = []
+        y_token_ids_with_blanks = [BLANK_ID]
+        y_tokens = []
+        y_tokens_with_blanks = [BLANK_TOKEN]
+        token_info = [{"text": BLANK_TOKEN, "s_start": 0, "s_end": 0,}]
+        word_info = []
+        segment_info = []
+
+        segment_s_pointer = 1
+        word_s_pointer = 1
+
+        for segment in segments:
```

Comment:
Lots of code, to make life easier in long run, please document inline as much as possible all steps and why they are done / what they are doing.